crates/wasm-mutate/src/mutators/modify_init_exprs.rs

```diff
@@ -0,0 +1,190 @@
+//! This mutator modifies the constant initializer expressions between various valid forms in
+//! entities which require constant initializers.
+
+use crate::mutators::translate::{self, InitExprContext, Item, Translator};
+use crate::{Error, Mutator, Result};
+
+use rand::Rng;
+use wasm_encoder::{ElementSection, GlobalSection, Instruction};
+use wasmparser::{ElementSectionReader, GlobalSectionReader, InitExpr, Operator, Type};
+
+#[derive(Copy, Clone)]
+pub enum InitExpressionMutator {
+    Global,
+    ElementOffset,
+    ElementFunc,
+}
+
+struct InitTranslator {
+    new_value: u128,
+    skip_inits: u32,
+    context: InitExprContext,
+}
+
+impl Translator for InitTranslator {
+    fn as_obj(&mut self) -> &mut dyn Translator {
+        self
+    }
+
+    fn remap(&mut self, item: Item, idx: u32) -> Result<u32> {
+        match (self.context, item) {
+            // Element's function items can either be presented to `Translator` via `remap` here or
+            // via expressions (the function below). The API here is insufficiently flexible to
+            // produce null funcrefs so at least lets try pointing at the lowest numbered function
+            // instead.
+            //
+            // TODO: may make sense to actually generate the functions pointed at more randomly.
+            (InitExprContext::ElementFunction, Item::Function) => {
+                if self.skip_inits != 0 {
+                    self.skip_inits -= 1;
+                    Ok(idx)
+                } else {
+                    log::trace!("... replacing referenced function index with 0");
+                    self.skip_inits = u32::MAX;
+                    Ok(0)
+                }
+            }
+            _ => Ok(idx),
+        }
+    }
```

Comment:
Shouldn't `remap` never get called for the to-be-mutated entity inside an init expression since you override `tanslate_init_expr` (and never call `remap` inside it for the target entity) below?

Reply:
The `Mutator::mutate` calls `Translator::translate_element` which calls `translate::element` which in turn can call either `Translator::remap` or `Translator::translate_init_expr` depending on what shape the `elem`'s initializers have.

Reply:
Ah okay -- do you want to override `translate::element` then so that you can produce null funcrefs there? Also fine to just stick with this as is for now.

Reply:
I'll investigate this, yeah. I wasn't sure initially if `ElementItem::Func` style elements can actually encode nulls and I'm also not sure if encoding between `ElementItem::Func` and `ElementItem::Expr` is meaningfully different – if it is, I think there's likely a fair amount of value in not retaining `ElementItem::Func` encoding over producing `ref.null`s.

Reply:
Ah that's a good point!

I think actually what you have now is good then, because the default table elements are null, so we can simplify `ElementItem::Func`s to null via removing elements from the segment.

But we want to generate/support/mutate both kinds of segment encodings just to exercise those code paths in parsers/runtimes/etc.